### ISSUE TITLE
False positive with \Ds\Vector::merge()

### ISSUE THREAD

**Issue by enumag** [vimeo/psalm#3515]
No idea what's wrong here...:

```php
<?php declare(strict_types = 1);

namespace App;

use DateTimeImmutable;
use Ds\Vector;

abstract class Z
{
	public function hello(): void
	{
		/** @var Vector<DateTimeImmutable> $vector */
		$vector = new Vector();
		$vector->merge($this->get());
	}

	/**
	 * @return Vector<DateTimeImmutable>
	 */
	abstract protected function get(): Vector;
}
```

```
ERROR: InvalidArgument - src/Z.php:14:18 - Argument 1 of Ds\Vector::merge expects iterable<mixed, TValue2:fn-ds\sequence::merge as mixed>, Ds\Vector<DateTimeImmutable> provided (see https://psalm.dev/004)
                $vector->merge($this->get());

```

**Comment by weirdan** [comment]
Reproduced more succinctly: https://psalm.dev/r/836ac33319

**Comment by muglug** [comment]
@enumag I think all these issues can just be fixed by adding the interface method signature to the implementing class. It's a bit of duplication, but it's necessary for Psalm not to be wildly confused.

**Comment by enumag** [comment]
@muglug Do you know which methods need that? I doubt it's really necessary for all of them...

**Comment by muglug** [comment]
I think any method that has a templated return type
